IGNITE-12824 .NET: Add BinaryConfiguration.TimestampConverter

nizhikov opened the pull request:
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

ptupitsyn commented on modules/platforms/dotnet/Apache.Ignite.Core/Binary/BinaryConfiguration.cs:
```diff
@@ -136,6 +137,11 @@ public BinaryConfiguration(params Type[] binaryTypes)
         /// </summary>
         public IBinarySerializer Serializer { get; set; }
 
+        /// <summary>
+        /// Default date time converter.
+        /// </summary>
```
```
        /// <summary>
        /// Gets or sets a converter between <see cref="DateTime"/> and Java Timestamp.
        /// Called from <see cref="IBinaryWriter.WriteTimestamp"/>, <see cref="IBinaryWriter.WriteTimestampArray"/>,
        /// <see cref="IBinaryReader.ReadTimestamp"/>, <see cref="IBinaryReader.ReadTimestampArray"/>.
        /// <para />
        /// See also <see cref="ForceTimestamp"/>.
        /// </summary>
```

ptupitsyn replied:
Also please update the doc for `ForceTimestamp` below like this:
```
Only UTC values are supported in Timestamp format. Other values will cause an exception on write, unless <see cref="TimestampConverter"/> is provided.
```

ptupitsyn reviewed (changes requested):
Looks good to me in general; please check some comments above, and make sure to add more tests in `BinaryDateTimeTest`.

nizhikov commented on modules/platforms/dotnet/Apache.Ignite.Core/Binary/IDateTimeConverter.cs:
```diff
@@ -0,0 +1,38 @@
+    /// <summary>
+    /// Converter for DateTime objects.
+    /// </summary>
+    public interface IDateTimeConverter
```
Renamed

nizhikov commented on modules/platforms/dotnet/Apache.Ignite.Core/Binary/BinaryConfiguration.cs:
```diff
@@ -136,6 +137,11 @@ public BinaryConfiguration(params Type[] binaryTypes)
+        /// <summary>
+        /// Default date time converter.
+        /// </summary>
+        public IDateTimeConverter DateTimeConverter { get; set; }
```
Renamed

nizhikov commented:
@ptupitsyn all comments fixed. Please, take a look one more time.

ptupitsyn commented on modules/platforms/dotnet/Apache.Ignite.Core.Tests/Binary/BinaryDateTimeTest.cs:
```diff
@@ -110,6 +111,34 @@ public void TestClassAttributes()
             AssertTimestampField<DateTimeClassAttribute2>((o, d) => o.Value = d, o => o.Value, "Value");
         }
 
+#if NETCOREAPP
```
I think this test will work on .NET Framework too (and we should have a test that works on .NET Framework).

nizhikov replied:
Fixed.

ptupitsyn commented on modules/platforms/dotnet/Apache.Ignite.Core.Tests/Binary/BinaryDateTimeTest.cs:
```diff
@@ -110,6 +111,34 @@ public void TestClassAttributes()
+            AssertTimestampField<DateTimeObj2>((o, d) => o.Value = d, o => o.Value, "Value");
+
+            var ex = Assert.Throws<BinaryObjectException>(() => binary.ToBinary<IBinaryObject>(new DateTime(1997, 8, 29)), 
+                "Converter Error!");
```
1. Unused variable `ex`.
2. `"Converter Error!"` is a message that will be displayed on assertion failure. I think the intent here is to check the exception message instead. To do that, add `Assert.AreEqual("Converter Error!", ex.Message)`.

nizhikov replied:
Thanks! Fixed

ptupitsyn reviewed (changes requested):
Please add tests to verify that converter is called in all possible cases:
* `DateTime` field write/read
* `DateTime[]` field write/read
* `DateTime` cache key write/read
* `DateTime[]` cache key write/read
* `DateTime` cache value write/read
* `DateTime[]` cache value write/read

nizhikov commented:
@ptupitsyn 

> Please add tests to verify that converter is called in all possible cases:

Thanks! Good catch!

I've added tests and they showed some bugs in the current changes.
Fixed it.

Please, take a look.

nizhikov commented:
https://ci.ignite.apache.org/viewLog.html?buildId=5802444&buildTypeId=IgniteTests24Java8_RunAllNet - tests run

ptupitsyn commented on modules/platforms/dotnet/Apache.Ignite.Core.Tests/Binary/BinaryDateTimeTest.cs:
```diff
@@ -134,10 +133,22 @@ public void TestCustomTimestampConverter()
+            var dt1 = new DateTime(1997, 8, 29, 0, 0, 0, DateTimeKind.Utc);
+
+            var ex = Assert.Throws<BinaryObjectException>(() => binary.ToBinary<DateTime>(dt1));
+            Assert.AreEqual("ToJavaTicks Error!", ex.Message);
```
Let's extract a constant to avoid repeating `"ToJavaTicks Error!"` value.

nizhikov replied:
Fixed.

ptupitsyn reviewed (changes requested):
@nizhikov 

1. What about some tests for cache key/value scenario and cache key field/value field? For now we only test `IBinary.ToBinary`, which can follow a different code path.

2. `TestCustomTimestampConverter` does not cover successful use case when the value is returned from the converter
![image](https://user-images.githubusercontent.com/10922892/103028436-d1943b00-4568-11eb-8a33-46acae8944ef.png)

nizhikov commented:
@ptupitsyn 

> What about some tests for cache key/value scenario and cache key field/value field? For now we only test IBinary.ToBinary, which can follow a different code path.

Tests for cache added.

> TestCustomTimestampConverter does not cover successful use case when the value is returned from the converter

Fixed.

ptupitsyn commented:
@nizhikov I've added one more test that does not rely on exceptions and pushed to the branch (hope you don't mind). Just to make sure that success path is tested as well.

nizhikov commented:
@ptupitsyn I'm glad that you help me with the patch. Appreciate it!

Now, PR looks good? What do you think? :)

ptupitsyn reviewed (approved):
Looks good, thank you!
